pkg/queue/breaker.go

```diff
@@ -60,12 +66,91 @@ func (b *Breaker) Maybe(thunk func()) bool {
 	case b.pendingRequests <- t:
 		// Pending request has capacity.
 		// Wait for capacity in the active queue.
-		b.activeRequests <- t
+		b.sem.Acquire()
 		// Defer releasing capacity in the active and pending request queue.
-		defer func() { <-b.activeRequests; <-b.pendingRequests }()
+		defer func() { b.sem.Release(); <-b.pendingRequests }()
 		// Do the thing.
 		thunk()
 		// Report success
 		return true
 	}
 }
+
+// NewSemaphore creates a semaphore with the desired maximal and initial capacity
+func NewSemaphore(maxCapacity, initialCapacity int32) *Semaphore {
+	if initialCapacity < 0 || initialCapacity > maxCapacity {
+		panic(fmt.Sprintf("Initial capacity must be between 0 and maximal capacity. Got %v.", initialCapacity))
+	}
+	queue := make(chan token, maxCapacity)
+	sem := Semaphore{queue: queue}
+	if initialCapacity > 0 {
+		sem.AddCapacity(initialCapacity)
+	}
+	return &sem
+}
+
+// Semaphore is an implementation of a semaphore based on Go channels
+// The number of available tokens is the number of elements in the buffered channel
+type Semaphore struct {
+	queue    chan token
+	token    token
+	reducers int32
+	capacity int32
+	mux      sync.Mutex
+}
+
+// Acquire receives the token from the semaphore, potentially blocking
+func (s *Semaphore) Acquire() {
+	<-s.queue
+}
+
+// Release releases the token to the queue
+// The operation is potentially blocking when the queue is full
+func (s *Semaphore) Release() {
+	s.AddCapacity(1)
+}
+
+// ReduceCapacity removes tokens from the rotation
+// It tries to acquire as many tokens as possible, if there are not enough tokens in the queue,
+// it postpones the operation for the future by increasing the `reducers` counter
+func (s *Semaphore) ReduceCapacity(size int32) error {
+	s.mux.Lock()
+	if size > s.capacity {
+		return errors.New("the capacity that is released must be <= to added capacity")
+	}
+	defer s.mux.Unlock()
```

Comment:
Please move this line above the if, otherwise you can get infinite locking here.